lfs/transfer_queue.go

```diff
@@ -0,0 +1,212 @@
+				// If an API authorization has not occured, we wait until we're woken up.
+				q.authCond.L.Lock()
+				if atomic.LoadInt32(&q.clientAuthorized) == 0 {
+					q.authCond.Wait()
+				}
+				q.authCond.L.Unlock()
+
+				obj, err := t.Check()
+				if err != nil {
+					q.errorc <- err
+					wg.Done()
+					continue
+				}
+				if obj != nil {
+					q.wg.Add(1)
+					t.SetObject(obj)
+					q.transferc <- t
+				}
+				wg.Done()
+			}
+		}()
+	}
+
+	q.bar.Prefix(fmt.Sprintf("(%d of %d files) ", q.finished, len(q.transferables)))
+	q.bar.Start()
+
+	for _, t := range q.transferables {
+		wg.Add(1)
+		apic <- t
+	}
+
+	<-workersReady
+	q.authCond.Signal() // Signal the first goroutine to run
+	close(apic)
+	wg.Wait()
+
+	close(q.transferc)
+}
+
+// processBatch processes the queue of transfers using the batch endpoint,
+// making only one POST call for all objects. The results are then handed
+// off to the transfer workers.
+func (q *TransferQueue) processBatch() {
+	q.files = 0
+	transfers := make([]*objectResource, 0, len(q.transferables))
+	for _, t := range q.transferables {
+		transfers = append(transfers, &objectResource{Oid: t.Oid(), Size: t.Size()})
+	}
+
+	objects, err := Batch(transfers)
+	if err != nil {
+		q.errorc <- err
+		sendApiEvent(apiEventFail)
+		return
+	}
+
+	for _, o := range objects {
+		if _, ok := o.Links[q.transferKind]; ok {
+			// This object needs to be transfered
+			if transfer, ok := q.transferables[o.Oid]; ok {
+				q.files++
+				q.wg.Add(1)
+				transfer.SetObject(o)
+				q.transferc <- transfer
+			}
+		}
+	}
+
+	close(q.transferc)
+	q.bar.Prefix(fmt.Sprintf("(%d of %d files) ", q.finished, q.files))
+	q.bar.Start()
+	sendApiEvent(apiEventSuccess) // Wake up transfer workers
+}
+
+// Process starts the transfer queue and displays a progress bar. Process will
+// do individual or batch transfers depending on the Config.BatchTransfer() value.
+// Process will transfer files sequentially or concurrently depending on the
+// Concig.ConcurrentTransfers() value.
+func (q *TransferQueue) Process() {
+	q.bar = pb.New64(q.size)
+	q.bar.SetUnits(pb.U_BYTES)
+	q.bar.ShowBar = false
+
+	// This goroutine collects errors returned from transfers
+	go func() {
+		for err := range q.errorc {
+			q.errors = append(q.errors, err)
+		}
+	}()
+
+	// This goroutine watches for apiEvents. In order to prevent multiple
+	// credential requests from happening, the queue is processed sequentially
+	// until an API request succeeds (meaning authenication has happened successfully).
+	// Once the an API request succeeds, all worker goroutines are woken up and allowed
+	// to process transfers. Once a success happens, this goroutine exits.
+	go func() {
+		for {
+			event := <-apiEvent
+			switch event {
+			case apiEventSuccess:
+				atomic.StoreInt32(&q.clientAuthorized, 1)
+				q.authCond.Broadcast() // Wake all remaining goroutines
+				return
+			case apiEventFail:
+				q.authCond.Signal() // Wake the next goroutine
+			}
+		}
+	}()
```

Comment:
Does it make sense to add an `apiEventShutdown` so that a TransferQueue can clean up any lingering goroutines once it's done? I don't think we have any cases where multiple transfer queues are processed in the same process, so it may not matter.